test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs

```diff
@@ -599,20 +616,20 @@ public void SdcaL1UpdateSUTest(int test)
 
             CpuMathUtils.SdcaL1UpdateSparse(DEFAULT_SCALE, src.Length, src, idx, idx.Length, DEFAULT_SCALE, v, w);
             var actual = w;
-            Assert.Equal(expected, actual, comparer);
+            Assert.Equal(expected, actual, _comparer);
         }
     }
 
     internal class FloatEqualityComparer : IEqualityComparer<float>
     {
         public bool Equals(float a, float b)
         {
-            return Math.Abs(a - b) < 1e-5f;
+            return Math.Abs(a - b) < 1e-3f;
```

Comment:
Why is this necessary?  I think we should leave the more precise version here.

Reply:
When it was 1e-5f or 1e-4f, unfortunately one of the unit tests failed even though the difference between the floats appeared 0.0001.  Should I leave this as is or change it in some way?

Reply:
IMO, it is unfortunate to degrade the precision checking of all the other tests just because it didn't work for 1 unit test.  I'd prefer to leave this as `return Math.Abs(a - b) < 1e-5f;` and investigate the one unit test that it doesn't work for, and figure out some other way for that test to work.

Reply:
Agree!  I re-ran the unit tests, and found out 8 unit tests (output attached below) failed after switching it back to `1e-5f`, but they are all tests for matrix multiplication.  I also want to use `1e-5f`, so should I (1) change those expected values, e.g. from `-416.68` to `-416.6801`, or (2) change my inputs to those MatMul unit tests, or (3) declare a separate comparer class for those MatMul unit tests?
```
[xUnit.net 00:00:00.5643501]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [-416.68, -415.68, -414.68, -413.68, -412.68, ...]) [FAIL]
Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [-416.68, -415.68, -414.68, -413.68, -412.68, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [-416.68, -415.68, -414.68, -413.68, -412.68, ...]
Actual:   Single[] [-416.6801, -415.6801, -414.6801, -413.6801, -412.6801, ...]

[xUnit.net 00:00:00.6144601]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulPAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [38.25, 39.25, 40.25, 41.25, 42.25, ...]) [FAIL]
[xUnit.net 00:00:00.6151858]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [70.56, -84.68, -349.36, 501.24, -3825.32, ...]) [FAIL]
Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulPAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [38.25, 39.25, 40.25, 41.25, 42.25, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [38.25, 39.25, 40.25, 41.25, 42.25, ...]
Actual:   Single[] [38.25002, 39.25002, 40.25002, 41.25002, 42.25002, ...]

Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [70.56, -84.68, -349.36, 501.24, -3825.32, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [70.56, -84.68, -349.36, 501.24, -3825.32, ...]
Actual:   Single[] [70.56001, -84.68, -349.36, 501.24, -3825.32, ...]

[xUnit.net 00:00:00.6163750]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [-416.68, -416.68, -416.68, -416.68, -416.68, ...]) [FAIL]
Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [-416.68, -416.68, -416.68, -416.68, -416.68, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [-416.68, -416.68, -416.68, -416.68, -416.68, ...]
Actual:   Single[] [-416.6801, -416.6801, -416.6801, -416.6801, -416.6801, ...]

[xUnit.net 00:00:00.6171754]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [70.56, -85.68, -351.36, 498.24, -3829.32, ...]) [FAIL]
[xUnit.net 00:00:00.6173835]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranPAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [33.32, -39.46, -163.92, 238.28, -1804.29, ...]) [FAIL]
Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [70.56, -85.68, -351.36, 498.24, -3829.32, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [70.56, -85.68, -351.36, 498.24, -3829.32, ...]
Actual:   Single[] [70.56001, -85.68, -351.36, 498.24, -3829.32, ...]

Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranPAAddTest(matTest: 0, srcTest: 0, dstTest: 0, expected: [33.32, -39.46, -163.92, 238.28, -1804.29, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [33.32, -39.46, -163.92, 238.28, -1804.29, ...]
Actual:   Single[] [33.32, -39.46, -163.92, 238.28, -1804.29, ...]

[xUnit.net 00:00:00.6190299]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulPATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [38.25, 38.25, 38.25, 38.25, 38.25, ...]) [FAIL]
[xUnit.net 00:00:00.6215487]     Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranPATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [33.32, -40.46, -165.92, 235.28, -1808.29, ...]) [FAIL]
Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulPATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [38.25, 38.25, 38.25, 38.25, 38.25, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [38.25, 38.25, 38.25, 38.25, 38.25, ...]
Actual:   Single[] [38.25002, 38.25002, 38.25002, 38.25002, 38.25002, ...]

Failed   Microsoft.ML.CpuMath.UnitTests.CpuMathUtilsUnitTests.MatMulTranPATest(matTest: 0, srcTest: 0, dstTest: 0, expected: [33.32, -40.46, -165.92, 235.28, -1808.29, ...])
Error Message:
 Assert.Equal() Failure
Expected: Single[] [33.32, -40.46, -165.92, 235.28, -1808.29, ...]
Actual:   Single[] [33.32, -40.46, -165.92, 235.28, -1808.29, ...]
```

Reply:
I guess I'd change the expected values if that works.

Reply:
Update: After I have changed the expected values to be the face values of the actual values (I used `ITestOutputHelper output` described in https://xunit.github.io/docs/capturing-output.html), 6 unit tests still failed with `< 1e-5f`.  2 unit tests failed with `< 1e-4f`. All unit tests passed with `< 1e-3f`.

Should I declare another comparer class with `< 1e-3f` just for the MatMul?

Reply:
I have added another comparer class with `< 1e-3f` specifically for `MatMul` in my latest commit below.  Please feel free to comment if I should improve it in any way.